File: app/src/main/java/com/mnj/java_solutions/slidingwindow/SubarraySumsDivisibleByKLC974.java
package com.mnj.java_solutions.slidingwindow;

import java.util.HashMap;

public class SubarraySumsDivisibleByKLC974 {

    public void subarraysDivByK() {
        int[] arr = {4, 5, 0, -2, -3, 1};
        int k = 5;

        int j = 0;
        int sum = 0;
        int rem = 0;
        int count = 0;

        HashMap<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);

        while (j < arr.length) {
            sum = sum + arr[j];

            rem = sum % k;

            if (rem < 0)
                rem = rem + k;

            if (map.containsKey(rem)) {
                count = count + map.get(rem);
                map.put(rem, map.getOrDefault(rem, 0) + 1);
            } else {
                map.put(rem, 1);
            }
            j++;
        }

        System.out.println("==>> Sub arrays sum divisible by K:" + count);
    }
}

File: app/src/main/java/com/mnj/java_solutions/slidingwindow/SubarraySumsDivisibleByKLC974.kt
package com.mnj.java_solutions.slidingwindow

class SubarraySumsDivisibleByKLC974 {
    fun subarraysDivByK() {
        val arr = intArrayOf(4, 5, 0, -2, -3, 1)
        val k = 5
        var j = 0
        var sum = 0
        var rem = 0
        var count = 0
        val map = HashMap<Int, Int>()
        map[0] = 1
        while (j < arr.size) {
            sum = sum + arr[j]
            rem = sum % k
            if (rem < 0) rem = rem + k
            if (map.containsKey(rem)) {
                count = count + map[rem]!!
                map[rem] = map.getOrDefault(rem, 0) + 1
            } else {
                map[rem] = 1
            }
            j++
        }
        println("==>> Sub arrays sum divisible by K:$count")
    }
}
